x-pack/plugin/mapper-exponential-histogram/src/main/java/org/elasticsearch/xpack/exponentialhistogram/EncodedHistogramData.java

```diff
@@ -0,0 +1,237 @@
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.ByteArrayStreamInput;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.exponentialhistogram.ExponentialHistogram.MAX_SCALE;
+import static org.elasticsearch.exponentialhistogram.ExponentialHistogram.MIN_SCALE;
+
+/**
+ * Encodes the data of an exponential histogram in a compact format as byte array.
+ * Note that some data of exponential histograms is stored outside of this array as separate doc values for better compression,
+ * e.g. the zero threshold. This data is therefore not part of this encoding.
+ **/
+final class EncodedHistogramData {
+
+    /*
+     The encoding has the following format:
+        - 1 byte: scale + flags
+                  scale has a range of less than 64, so we can use the two remaining bits for flags
+                  Currently the only flag is HAS_NEGATIVE_BUCKETS_FLAG, which indicates that there are negative buckets
+                  If this flag is not set, we can skip encoding the length of the negative buckets and therefore save a
+                  bit of space for this typical case.
+        - VInt: (optional) length of encoded negative buckets in bytes, if HAS_NEGATIVE_BUCKETS_FLAG is set
+        - byte[]: encoded negative buckets, details on the encoding below
+        - byte[]: encoded positive buckets, details on the encoding below
+     There is no end marker for the encoded buckets, therefore the total length of the encoded data needs to be known when decoding
+
+     The following scheme is used to encode the negative and positive buckets:
+        - if there are no buckets, the result is an empty array (byte[0])
+        - write the index of the first bucket as ZigZag-VLong
+        - write the count of the first bucket as ZigZag-VLong
+        - for each remaining bucket:
+           - if the index of the bucket is exactly previousBucketIndex+1, write the count for the bucket as ZigZag-VLong
+           - Otherwise there is at least one empty bucket between this one and the previous one.
+             We compute that number as n=currentBucketIndex-previousIndex-1 and then write -n out as
+             ZigZag-VLong followed by the count for the bucket as ZigZag-VLong. The negation is performed to allow to
+             distinguish the cases when decoding.
+
+     While this encoding is designed for sparse histograms, it compresses well for dense histograms too.
+     For fully dense histograms it effectively results in encoding the index of the first bucket, followed by just an array of counts.
+     For sparse histograms it corresponds to an interleaved encoding of the bucket indices with delta compression and the bucket counts.
```

Comment:
Have you tried compressing the bucket counts as deltas or delta of deltas, too? I wonder if that could reduce the storage.

Reply:
If it sounds too complex for now, we could also do this another time, and create a new encoding version for it. Although, I think it would be good to get a rough estimation now. That wouldn't require a full implementation.

Reply:
Hm, I just realized that this probably wouldn't work well because the delta could be negative and confused with empty buckets.

Reply:
I did a quick test. As expected even when having 1000+ points per histogram, the gains in terms of storage are less than 0.1%. I think it would be best to first try out encodings allowing sub 1 byte per value, ideally evaluating it on some real data set. But let's do this investiagation separately from this PR.

Reply:
> I think it would be best to first try out encodings allowing sub 1 byte per value, ideally evaluating it on some real data set. But let's do this investiagation separately from this PR.

++